File: apps/api/src/modules/llm/cache/semantic-cache.service.ts
/**
 * 语义缓存服务（P0-2）
 *
 * 作用：相同语义的提问 → 复用之前的 LLM 回复，避免重复推理
 *
 * 关键设计：
 *  - embedding 走 Qwen text-embedding-v3（v13 已有 OpenAI 兼容 client，零新增依赖）
 *  - 向量存 Qdrant（payload 含 role / userId / cacheType / 原文 hash）
 *  - 相似度阈值 0.92（白名单启用：interview_question / general_qa）
 *  - 评分题强制 miss（payload.cacheType === 'scoring' 跳过）
 *  - 异步 set：不阻塞主调用路径
 *  - 埋点：hit / miss 计数走 SessionCostTracker
 *
 * Redis 角色：仍然存"精确 cacheKey 命中"的快速层（Qdrant 查询前先 hash）
 */

import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import OpenAI from 'openai';
import { randomUUID } from 'crypto';
import { QdrantService } from '../../../infra/qdrant/qdrant.service';
import { RedisService } from '../../../infra/redis/redis.service';
import { fnv1a } from './prompt-cache.strategy';

export type SemanticCacheType =
  | 'interview_question' // ✅ 启用（白名单）
  | 'general_qa' // ✅ 启用
  | 'scoring' // ❌ 强制 miss
  | 'tool_result' // ❌ 强制 miss（工具结果不该缓存）
  | 'resume_parse' // ❌ 强制 miss（个性化强）
  | 'report_generate'; // ❌ 强制 miss

export interface SemanticCacheHit {
  hit: true;
  cachedResponse: string;
  similarity: number;
  cacheId: string;
}

export interface SemanticCacheMiss {
  hit: false;
  reason: 'disabled' | 'whitelist' | 'low_similarity' | 'cold';
}

export type SemanticCacheResult = SemanticCacheHit | SemanticCacheMiss;

export interface SemanticCacheSetParams {
  userId: string;
  cacheType: SemanticCacheType;
  query: string;
  response: string;
  metadata?: Record<string, any>;
}

export interface SemanticCacheLookupParams {
  userId: string;
  cacheType: SemanticCacheType;
  query: string;
  /** 命中相似度阈值，默认 0.92 */
  threshold?: number;
}

const COLLECTION = 'semantic_cache';
const VECTOR_SIZE = 1024; // Qwen text-embedding-v3 默认 1024 维
const REDIS_HASH_PREFIX = 'sc:hash:'; // exact-match fast path
const REDIS_TTL_SECONDS = 3600;

@Injectable()
export class SemanticCacheService implements OnModuleInit {
  private readonly logger = new Logger(SemanticCacheService.name);
  private embedder: OpenAI;
  private enabled: boolean;
  private whitelist: Set<SemanticCacheType>;

  /** Provider 无关的强制 miss 类型（涉及个性化、评估、副作用） */
  private readonly blacklist = new Set<SemanticCacheType>([
    'scoring',
    'tool_result',
    'resume_parse',
    'report_generate',
  ]);

  constructor(
    private config: ConfigService,
    private qdrant: QdrantService,
    private redis: RedisService,
  ) {}

  async onModuleInit() {
    this.enabled = this.config.get<string>('semanticCache.enabled') !== 'false';
    const whitelistCfg =
      this.config.get<string>('semanticCache.whitelist') || 'interview_question,general_qa';
    this.whitelist = new Set(whitelistCfg.split(',').map((s) => s.trim()) as SemanticCacheType[]);

    // 复用 Qwen 客户端做 embedding
    this.embedder = new OpenAI({
      apiKey: this.config.get<string>('qwen.apiKey'),
      baseURL: this.config.get<string>('qwen.baseUrl'),
    });

    if (this.enabled) {
      await this.ensureCollection();
    }
    this.logger.log(
      `✅ SemanticCache ready (enabled=${this.enabled}, whitelist=${[...this.whitelist].join(',')})`,
    );
  }

  private async ensureCollection() {
    try {
      const client = this.qdrant.getClient();
      const collections = await client.getCollections();
      const exists = collections.collections?.some((c) => c.name === COLLECTION);
      if (!exists) {
        await client.createCollection(COLLECTION, {
          vectors: { size: VECTOR_SIZE, distance: 'Cosine' },
        });
        // 建索引
        await client.createPayloadIndex(COLLECTION, {
          field_name: 'userId',
          field_schema: 'keyword',
        });
        await client.createPayloadIndex(COLLECTION, {
          field_name: 'cacheType',
          field_schema: 'keyword',
        });
        this.logger.log(`Created Qdrant collection ${COLLECTION}`);
      }
    } catch (err: any) {
      this.logger.warn(`Qdrant collection init failed: ${err.message}`);
    }
  }

  /**
   * 查缓存
   * 1. 黑名单直接 miss
   * 2. 不在白名单直接 miss
   * 3. 先查 Redis hash 精确桶
   * 4. 桶未命中则 embedding + Qdrant cosine
   */
  async lookup(params: SemanticCacheLookupParams): Promise<SemanticCacheResult> {
    if (!this.enabled) return { hit: false, reason: 'disabled' };
    if (this.blacklist.has(params.cacheType)) return { hit: false, reason: 'whitelist' };
    if (!this.whitelist.has(params.cacheType)) return { hit: false, reason: 'whitelist' };

    // Fast path:精确 hash
    const fastKey = this.fastKey(params.userId, params.cacheType, params.query);
    try {
      const exact = await this.redis.get(fastKey);
      if (exact) {
        const parsed = JSON.parse(exact) as { response: string; cacheId: string };
        return { hit: true, cachedResponse: parsed.response, similarity: 1.0, cacheId: parsed.cacheId };
      }
    } catch (e) {
      // Redis 故障不阻塞主流程
    }

    // Slow path:embedding 检索
    let vector: number[];
    try {
      vector = await this.embed(params.query);
    } catch (err: any) {
      this.logger.warn(`embedding failed, miss: ${err.message}`);
      return { hit: false, reason: 'cold' };
    }

    try {
      const client = this.qdrant.getClient();
      const threshold = params.threshold ?? 0.92;
      const search = await client.search(COLLECTION, {
        vector,
        limit: 1,
        score_threshold: threshold,
        filter: {
          must: [
            { key: 'userId', match: { value: params.userId } },
            { key: 'cacheType', match: { value: params.cacheType } },
          ],
        },
        with_payload: true,
      });

      if (search.length === 0) return { hit: false, reason: 'low_similarity' };

      const top = search[0];
      const payload = top.payload as { response: string; cacheId: string };
      return {
        hit: true,
        cachedResponse: payload.response,
        similarity: top.score,
        cacheId: payload.cacheId,
      };
    } catch (err: any) {
      this.logger.warn(`Qdrant search failed, miss: ${err.message}`);
      return { hit: false, reason: 'cold' };
    }
  }

  /**
   * 写缓存 - 异步，不阻塞主调用
   * 失败也不抛（缓存是优化路径，不应影响主流程）
   */
  setAsync(params: SemanticCacheSetParams): void {
    if (!this.enabled) return;
    if (this.blacklist.has(params.cacheType)) return;
    if (!this.whitelist.has(params.cacheType)) return;

    // 异步：setImmediate 让出当前 tick
    setImmediate(() => {
      this.setInternal(params).catch((err) => {
        this.logger.debug(`semantic-cache set failed (silent): ${err.message}`);
      });
    });
  }

  private async setInternal(params: SemanticCacheSetParams): Promise<void> {
    const vector = await this.embed(params.query);
    const cacheId = randomUUID();
    const client = this.qdrant.getClient();
    await client.upsert(COLLECTION, {
      wait: false,
      points: [
        {
          id: cacheId,
          vector,
          payload: {
            userId: params.userId,
            cacheType: params.cacheType,
            query: params.query,
            response: params.response,
            createdAt: Date.now(),
            ...params.metadata,
          },
        },
      ],
    });
    // Redis 精确层
    const fastKey = this.fastKey(params.userId, params.cacheType, params.query);
    await this.redis.set(
      fastKey,
      JSON.stringify({ response: params.response, cacheId }),
      REDIS_TTL_SECONDS,
    );
  }

  private async embed(text: string): Promise<number[]> {
    const res = await this.embedder.embeddings.create({
      model: 'text-embedding-v3',
      input: text.slice(0, 2048),
      encoding_format: 'float',
      dimensions: VECTOR_SIZE,
    } as any);
    return res.data[0].embedding as number[];
  }

  private fastKey(userId: string, cacheType: SemanticCacheType, query: string): string {
    const hash = fnv1a(`${userId}::${cacheType}::${query.trim().toLowerCase()}`).toString(16);
    return `${REDIS_HASH_PREFIX}${cacheType}:${userId}:${hash}`;
  }
}

File: apps/py/src/interview_agent/modules/llm/cache/semantic_cache_service.py
"""Semantic Cache — 与 NestJS SemanticCacheService 像素级对齐。

行为：
- 黑名单（scoring / tool_result / resume_parse / report_generate）强制 miss
- 白名单（interview_question / general_qa）启用
- Fast path：Redis hash 精确桶（< 5ms）
- Slow path：Qwen text-embedding-v3 + Qdrant cosine（阈值 0.92）
- setImmediate-style 异步写入（不阻塞主调用）

降级策略：
- 无 QWEN_API_KEY / Qdrant 不可用时，全部走 Redis 精确层
"""
import asyncio
import json
import logging
import os
from typing import Literal

from interview_agent.infra.redis_client import get_redis
from interview_agent.modules.llm.cache.prompt_cache_strategy import fnv1a

logger = logging.getLogger(__name__)

COLLECTION = "semantic_cache"
VECTOR_SIZE = 1024
REDIS_HASH_PREFIX = "sc:hash:"
REDIS_TTL_SECONDS = 3600

SemanticCacheType = Literal[
    "interview_question",
    "general_qa",
    "scoring",
    "tool_result",
    "resume_parse",
    "report_generate",
]

BLACKLIST = {"scoring", "tool_result", "resume_parse", "report_generate"}
WHITELIST = {"interview_question", "general_qa"}


class SemanticCacheService:
    """语义缓存服务单例。

    用法：
    ```python
    svc = SemanticCacheService.instance()
    r = await svc.lookup(user_id, "interview_question", "什么是 React Fiber?")
    if r.hit:
        return r.cached_response
    # ... 调 LLM ...
    svc.set_async(user_id, "interview_question", query, response)
    ```
    """

    _instance: "SemanticCacheService | None" = None
    _qdrant = None  # lazy init
    _embedder = None  # lazy init

    def __init__(self):
        self.enabled = True
        self._qdrant_failed = False

    @classmethod
    def instance(cls) -> "SemanticCacheService":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def _fast_key(self, user_id: str, cache_type: str, query: str) -> str:
        h = format(fnv1a(f"{user_id}::{cache_type}::{query.strip().lower()}"), "x")
        return f"{REDIS_HASH_PREFIX}{cache_type}:{user_id}:{h}"

    async def lookup(
        self,
        user_id: str,
        cache_type: SemanticCacheType,
        query: str,
        threshold: float = 0.92,
    ) -> dict:
        """查缓存：返回 {hit, cachedResponse, similarity, cacheId} 或 {hit:false, reason}。"""
        if not self.enabled:
            return {"hit": False, "reason": "disabled"}
        if cache_type in BLACKLIST:
            return {"hit": False, "reason": "whitelist"}
        if cache_type not in WHITELIST:
            return {"hit": False, "reason": "whitelist"}

        # Fast path：Redis 精确
        try:
            redis = get_redis()
            exact = await redis.get(self._fast_key(user_id, cache_type, query))
            if exact:
                data = json.loads(exact)
                return {
                    "hit": True,
                    "cachedResponse": data["response"],
                    "similarity": 1.0,
                    "cacheId": data["cacheId"],
                }
        except Exception as e:
            logger.warning(f"semantic-cache fast path error: {e}")

        # Slow path：embedding + Qdrant
        try:
            vector = await self._embed(query)
            results = await self._qdrant_search(vector, user_id, cache_type, threshold)
            if results:
                top = results[0]
                return {
                    "hit": True,
                    "cachedResponse": top["payload"]["response"],
                    "similarity": top["score"],
                    "cacheId": top["id"],
                }
            return {"hit": False, "reason": "low_similarity"}
        except Exception as e:
            logger.warning(f"semantic-cache slow path error: {e}")
            return {"hit": False, "reason": "cold"}

    def set_async(
        self,
        user_id: str,
        cache_type: SemanticCacheType,
        query: str,
        response: str,
        metadata: dict | None = None,
    ) -> None:
        """异步写缓存（不阻塞主流程）。"""
        if not self.enabled or cache_type in BLACKLIST or cache_type not in WHITELIST:
            return
        # fire-and-forget
        asyncio.create_task(self._set_internal(user_id, cache_type, query, response, metadata))

    async def _set_internal(
        self, user_id, cache_type, query, response, metadata
    ) -> None:
        try:
            vector = await self._embed(query)
            cache_id = str(hash((user_id, query, response)) & 0x7FFFFFFFFFFFFFFF)
            await self._qdrant_upsert(cache_id, vector, {
                "userId": user_id,
                "cacheType": cache_type,
                "query": query,
                "response": response,
                "createdAt": 0,  # server 端时间
                **(metadata or {}),
            })
            redis = get_redis()
            await redis.set(
                self._fast_key(user_id, cache_type, query),
                json.dumps({"response": response, "cacheId": cache_id}),
                ex=REDIS_TTL_SECONDS,
            )
        except Exception as e:
            logger.debug(f"semantic-cache set silent fail: {e}")

    async def _embed(self, text: str) -> list[float]:
        """Qwen text-embedding-v3 → 1024 维向量。

        无 QWEN_API_KEY / OpenAI 包未装时返回 None（让 slow path 跳过）。
        """
        from interview_agent.config import settings

        if not settings.QWEN_API_KEY or "placeholder" in settings.QWEN_API_KEY:
            raise RuntimeError("embedding disabled: QWEN_API_KEY not set")
        try:
            from openai import AsyncOpenAI
        except ImportError:
            raise RuntimeError("openai package not installed")
        if self._embedder is None:
            self._embedder = AsyncOpenAI(
                api_key=settings.QWEN_API_KEY, base_url=settings.QWEN_BASE_URL
            )
        resp = await self._embedder.embeddings.create(
            model="text-embedding-v3",
            input=text[:2048],
            encoding_format="float",
            dimensions=VECTOR_SIZE,
        )
        return resp.data[0].embedding

    async def _qdrant_search(self, vector, user_id, cache_type, threshold) -> list[dict]:
        """Qdrant cosine search。无 qdrant-client 时直接返回 []（Redis 精确层兜底）。"""
        try:
            from qdrant_client import AsyncQdrantClient
            from interview_agent.config import settings
            if self._qdrant is None and not self._qdrant_failed:
                self._qdrant = AsyncQdrantClient(url=settings.QDRANT_URL)
            if self._qdrant is None:
                return []
            return await self._qdrant.search(
                collection_name=COLLECTION,
                query_vector=vector,
                limit=1,
                score_threshold=threshold,
                query_filter={
                    "must": [
                        {"key": "userId", "match": {"value": user_id}},
                        {"key": "cacheType", "match": {"value": cache_type}},
                    ]
                },
                with_payload=True,
            )
        except Exception as e:
            logger.debug(f"qdrant search skipped: {e}")
            self._qdrant_failed = True
            return []

    async def _qdrant_upsert(self, point_id, vector, payload) -> None:
        try:
            if self._qdrant is None:
                return
            await self._qdrant.upsert(
                collection_name=COLLECTION,
                points=[{"id": point_id, "vector": vector, "payload": payload}],
            )
        except Exception as e:
            logger.debug(f"qdrant upsert skipped: {e}")


def get_semantic_cache() -> SemanticCacheService:
    return SemanticCacheService.instance()